x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/Analyzer.java

```diff
@@ -799,4 +816,93 @@ private BitSet gatherPreAnalysisMetrics(LogicalPlan plan, BitSet b) {
         plan.forEachDown(p -> FeatureMetric.set(p, b));
         return b;
     }
+
+    private static class ImplicitCasting extends Rule<LogicalPlan, LogicalPlan> {
+
+        @Override
+        public LogicalPlan apply(LogicalPlan plan) {
+            return plan.transformExpressionsUp(EsqlScalarFunction.class, ImplicitCasting::cast);
+        }
+
+        private static Expression cast(EsqlScalarFunction f) {
+            List<Expression> args = f.arguments();
+            List<String[]> metaData = getMetaData(f);
+            List<Expression> newChildren = new ArrayList<>(args.size());
+            boolean childrenChanged = false;
+            List<DataType> supportedTypes = new ArrayList<>();
+            for (int i = 0; i < args.size(); i++) {
+                if (args.get(i).dataType() == KEYWORD && args.get(i).foldable()) {
+                    if (i < metaData.size()) {
+                        supportedTypes = getSupportedTypes(metaData.get(i));
+                    }
+                    if (supportedTypes.contains(KEYWORD) == false && supportedTypes.contains(TEXT) == false) {
+                        DataType desired = supportedTypes.contains(DATETIME) ? DATETIME
+                            : supportedTypes.contains(DOUBLE) ? DOUBLE
+                            : supportedTypes.contains(LONG) ? LONG
+                            : supportedTypes.contains(INTEGER) ? INTEGER
+                            : supportedTypes.contains(IP) ? IP
+                            : supportedTypes.contains(VERSION) ? VERSION
+                            : supportedTypes.contains(GEO_POINT) ? GEO_POINT
+                            : supportedTypes.contains(GEO_SHAPE) ? GEO_SHAPE
+                            : supportedTypes.contains(CARTESIAN_POINT) ? CARTESIAN_POINT
+                            : supportedTypes.contains(CARTESIAN_SHAPE) ? CARTESIAN_SHAPE
+                            : supportedTypes.contains(BOOLEAN) ? BOOLEAN
```

Comment:
Not sure what's the point of this? The supported types can be initialized through an array/map once and then use it for lookup.

Reply:
The supportedTypes here is a list of supported data types for one argument. If an input parameter is a string literal, it gets a list of supportedTypes for this argument, and find the desired data type to cast to, with priority listed. The intention of this piece of code is to find which data type to cast to.

Reply:
```DataType desired = supportedTypes(metadata.get(i)```

As a tangent:
I would expect moving forward that the multiple types expected by a functions to be compatible. Which means we can design the function to expect only a single type and have the conversion externalized similar to this rule.
Say a function expects an IP fields that could also be passed as a long.
The function would declared just the Ip datatype and we'd have a dedicated rule that for this specific function and parameter would convert the long to ip.

Reply:
Made some modifications after discussing with Costin, hopefully the updated version looks better.